Sources/Machine/abstract_machine.hpp

```diff
@@ -215,14 +215,15 @@ class AbstractMachine {
   void Save(const std::string &filename) const;
   void Save(std::ofstream &stream) const;
 
-  const AbstractHilbert &GetHilbert() const;
-
-  void SetHilbert(const AbstractHilbert &hilbert);
+  const AbstractHilbert &GetHilbert() const { return *hilbert_; };
```

Comment:
These should be implemented in the cc file?

Reply:
Also, at this point to be really consistent and avoid later complications you'd need to return `std::shared_ptr<const AbstractHilbert>`

Reply:
There is a difference between taking (joint) ownership of the Hilbert object and not taking it. Returning `shared_ptr` from here would cause every caller of `GetHilbert` to take ownership of the Hilbert object, at least temporarily. That does not always make sense. For instance, calling `GetHilbert.Size()` somewhere in the code does not need ownership of the Hilbert object, while `return GraphOperator(GetHilbert().shared_from_this(), ...)` does.

Reply:
But we can make `GetHilbert` return a `shared_ptr`. In that case, we can remove `enable_shared_from_this`. This would lead to a few more copies of `shared_ptr`s thyat aren't strictly necessary, but that probably would not make a noticeable difference.